File: src/ros2_bridge/mappers/geometry_msgs/pose_stamped.rs
//! Typed mapper for `geometry_msgs/msg/PoseStamped`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn pose_stamped_to_bus(msg: ros_env::geometry_msgs::msg::PoseStamped) -> crate::geometry_msgs::msg::v1::PoseStamped {
    crate::geometry_msgs::msg::v1::PoseStamped {
        header: Some(crate::ros2_bridge::mappers::std_msgs::header::header_to_bus(msg.header)),
        pose: Some(crate::ros2_bridge::mappers::geometry_msgs::pose::pose_to_bus(msg.pose)),
    }
}

pub(crate) fn pose_stamped_to_ros(bus: crate::geometry_msgs::msg::v1::PoseStamped) -> ros_env::geometry_msgs::msg::PoseStamped {
    ros_env::geometry_msgs::msg::PoseStamped {
        header: crate::ros2_bridge::mappers::std_msgs::header::header_to_ros(bus.header.unwrap_or_default()),
        pose: crate::ros2_bridge::mappers::geometry_msgs::pose::pose_to_ros(bus.pose.unwrap_or_default()),
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct GeometryMsgsPoseStampedMapper;

impl TypedTopicMapper for GeometryMsgsPoseStampedMapper {
    type Ros = ros_env::geometry_msgs::msg::PoseStamped;
    type Bus = crate::geometry_msgs::msg::v1::PoseStamped;

    fn type_name(&self) -> &'static str {
        "geometry_msgs/msg/PoseStamped"
    }

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(pose_stamped_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(pose_stamped_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/geometry_msgs/pose_stamped.py
"""Generated mapper for `geometry_msgs/msg/PoseStamped`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert
from robot_bus.ros2_bridge.mappers.std_msgs.header import header_to_bus, header_to_ros
from robot_bus.ros2_bridge.mappers.geometry_msgs.pose import pose_to_bus, pose_to_ros

_BusMsg = None


def _bus_cls():
    global _BusMsg
    if _BusMsg is None:
        from robot_bus.geometry_msgs.msg.v1 import PoseStamped as BusMsg

        _BusMsg = BusMsg
    return _BusMsg


def pose_stamped_to_bus(msg):
    BusMsg = _bus_cls()
    bus = BusMsg()
    bus.header.CopyFrom(header_to_bus(msg.header))
    bus.pose.CopyFrom(pose_to_bus(msg.pose))
    return bus


def pose_stamped_to_ros(bus):
    from geometry_msgs.msg import PoseStamped as RosMsg

    out = RosMsg()
    out.header = header_to_ros(bus.header)
    out.pose = pose_to_ros(bus.pose)
    return out


class GeometryMsgsPoseStampedMapper:
    _ros_type = None

    def ros_msg_type(self):
        cls = type(self)
        if cls._ros_type is None:
            from geometry_msgs.msg import PoseStamped as RosMsg

            cls._ros_type = RosMsg
        return cls._ros_type

    def ros_to_bus(self, msg) -> bytes:
        return pose_stamped_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        BusMsg = _bus_cls()
        bus = BusMsg()
        bus.ParseFromString(payload)
        return pose_stamped_to_ros(bus)
